---
title: "Assignment 9"
author: "Michael E. Mutersbaugh"
date: "November 8, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


##Batch Effects Exercise 1:

###Excercise Introduction: Load the admissions data from the dagdata package (which is available from the genomicsclass repository):
```{r}
library(devtools)
#install_github("genomicsclass/dagdata")
library(dagdata)
data(admissions)

#Familiarize yourself with this table:
admissions
```

###1: Let's compute the proportion of men who were accepted:
```{r}
index = which(admissions$Gender==1)
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100)
applied = sum(admissions$Number[index])
accepted/applied
```

### What is the proportion of women that were accepted?
```{r}
index = which(admissions$Gender==0)
accepted = sum(admissions$Number[index] * admissions$Percent[index]/100)
applied = sum(admissions$Number[index])
accepted/applied
```

#### From the two computations above, we can see that the proportion of men that were accepted was 0.4451951 and the proportion of women that were accepted was slightly lower at 0.3033351. 
  
###2: Now that we have observed different acceptance rates between genders, test for the significance of this result. If you perform an independence test, what is the p-value?
```{r}
#This difference actually led to a lawsuit. Now notice that looking at the data by major, the differences disappear.
admissions
```

### How can this be? This is referred to as Simpson's Paradox. In the following questions we will try to decipher why this is happening.
```{r}
#I created a table that has the totals for accepted and not-accepted by gender then used a chisq.test:
atab<-as.data.frame(matrix(0,2,2))
rownames(atab)<-c("Men","Women")
colnames(atab)<-c("Accepted","Rejected")
index = admissions$Gender==1
men=admissions[index,]
women=admissions[!index,]
atab[1,1]= sum(men$Number * men$Percent/100)
atab[1,2]= sum(men$Number*(1-men$Percent/100))
windex = which(admissions$Gender==0)
atab[2,1]= sum(women$Number * women$Percent/100)
atab[2,2]= sum(women$Number * (1-women$Percent/100))
atab
chisq.test(atab)$p.value

#Looking at the data based on major:
index = admissions$Gender==1
men = admissions[index,]
women = admissions[!index,]
print( data.frame( major=admissions[1:6,1],men=men[,3], women=women[,3]) )
```

#### From the first computation looking at the data based on gender, the p-value was 9.139492e-22. 

###3: We can quantify how "hard" a major is by using the percent of students that were accepted. Compute the percent that were accepted (regardless of gender) to each major and call this vector H. Which is the hardest major?
```{r}
major = admissions[1:6,1]
men = admissions[1:6,]
women =admissions[7:12,]
H = (men$Number*men$Percent/100 + women$Number*women$Percent/100) / (men$Number+women$Number)
H
major[which.min(H)]
```

#### From this, we can see that the hardest major is "F" out of the different levels. It is classified as the hardest major based on the percentage of students that were accepted (regardless of gender). Therefore, major F had the lowest proportion of students that were accepted and is hence the "hardest" major. 
  
###4: What proportion is accepted for this major?
```{r}
min(H)
```

#### From the table in question 3 and the computation above, the proportion accepted for this major is 0.06477591. 
  
###5: For men, what is the correlation between the number of applications across majors and H?
```{r}
cor(H,men$Number)
cor(men$Number,H) # same!
```

#### The correlation between the number of applications across majors and H, for men, is 0.7647567. 
  
###6: For women, what is the correlation between the number of applications across majors and H?
```{r}
cor(H,women$Number)
cor(women$Number,H) # same!
```

#### The correlation between the number of applications across majors and H, for women, is -0.6743393. 
  
###7: Given the answers to the above, which best explains the differences in admission percentages when we combine majors?
#### A) We made a coding mistake when computing the overall admissions percentages.
#### B) There were more total number of women applications which made the denominator much bigger.
#### C) There is confounding between gender and preference for "hard" majors: females are more likely to apply to harder majors.
#### D) The sample size for the individual majors was not large enough to draw the correct conclusion.

#### The answer that best explains the differences in admission percentages when we combine majors is C) There is a confounding between gender and preference for "hard" majors: females are more likely to apply to harder majors. This is because, when we computed the correlation between the number of applications across majors and H, for women we got a negative correlation, meaning that a decrease in application of majors is associated with an increase in applications of H. This is an inverse relationship and it is contrasted with the correlation calculated for men (which was positive). Hence, the correct answer is that there is a confounding between gender and preference, which essentially means that women are more likely to apply to harder majors than men.
  
##Batch Effects Exercise 2:

###Excercise Introduction: We will use the Bioconductor package Biobase which you can install with install_bioc function from rafalib:
```{r}
#Load the data for this gene expression dataset:
library(devtools)
install_github("genomicsclass/GSE5859")
library(Biobase)
library(GSE5859)
data(GSE5859)

#This is the original dataset from which we selected the subset used in GSE5859Subset. We can extract the gene expression data and sample information table using the Bioconductor functions exprs and pData like this:
geneExpression = exprs(e)
sampleInfo = pData(e)
```

###1: Familiarize yourself with the sampleInfo table. Note that some samples were processed at different times. This is an extraneous variable and should not affect the values in geneExpression. However, as we have seen in previous analyses, it does appear to have an effect so we will explore this here.
```{r}
#You can extract the year from each date like this:
year = format(sampleInfo$date,"%y")

#Note that ethnic group and year is almost perfectly confounded:
table(year,sampleInfo$ethnicity)
```
```{r}
#The number of unique years for which we have data is:
length(unique(year))
```

#### There are 5 unique years for which we have data. 
  
###2: For how many of these years do we have more than one ethnicity represented?
```{r}
tab = table(year, sampleInfo$ethnicity)
print(tab)
x = rowSums(tab != 0)
sum(x >= 2)
```

#### As seen in the computation above, there are 2 years for which there is more than one ethnicity represented. 
  
###3: Repeat the above exercise, but now, instead of year, consider the month as well. Specifically, instead of the year variable defined above use:
```{r}
month.year = format(sampleInfo$date,"%m%y")
```
### For what proportion of these month.year values do we have more than one ethnicity represented?
```{r}
mytab = table(month.year, sampleInfo$ethnicity)
print(mytab)
myx = rowSums(mytab != 0)
mean(myx >= 2)
```

#### As per the computation above, the proportion of these month.year values that there is more than one ethnicity represented is 0.04761905. This value implies that month.year and ethnicity are almost completely confounded, which means that it is hard to separate effects due to date from effects due to our outcome of interest.
  
###4: Perform a t-test (use rowttests) comparing CEU samples processed in 2002 to those processed in 2003. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values < 0.05?
```{r}
library(qvalue)
library(genefilter)
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("02","03") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

#### After performing a t-test, per the computation above, the number of genes with q-value less than 0.05 is 4308. 
  
###5: What is the estimate of pi0 provided by qvalue:
```{r}
qval$pi0
```

#### The estimate of pi0 provided by the qvalue is 0.3628642. It's important to note that the estimated percentage of genes that are differentially expressed is above 50%. This is one way to show the magnitude of the effect processing date has on the measurements.

###6: Now perform a t-test (use rowttests) comparing CEU samples processed in 2003 to those processed in 2004. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values less than 0.05?
```{r}
library(qvalue)
library(genefilter)
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("03","04") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

#### From the computation above, we can see that the number of genes with q-value less than 0.05 is 2463. From this we can confirm the processing date has an effect on our measurements.

###7: Now we are going to compare ethnicities as was done in the original publication in which these data were first presented. Use the qvalue function to compare the ASN population to the CEU population. Once again, use the qvalue function to obtain q-values. How many genes have q-values < 0.05?
```{r}
library(qvalue)
library(genefilter)
ethnicity = factor(sampleInfo$ethnicity)
index = which(ethnicity %in% c("CEU", "ASN"))
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

#### From the code above, we can see that the number of genes with q-values less than 0.05 is 7217. 
  
###8: Over 80% of genes are called differentially expressed between ethnic groups. However, due to the confounding with processing date, we need to confirm these differences are actually due to ethnicity. This will not be easy due to the almost perfect confounding. However, above we noted that two groups were represented in 2005. Just like we stratified by majors to remove the "major effect" in our admissions example, here we can stratify by year and perform a t-test comparing ASN and CEU, but only for samples processed in 2005. How many genes have q-values < 0.05 ?

###Notice the dramatic drop in the number of genes with q-value < 0.05 when we fix the year. However, the sample size is much smaller in this latest analysis which means we have less power:
  
``{r}
table(sampleInfo$ethnicity[index])

How many genes have q-values < 0.05?:
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
index = which(ethnicity%in% c("CEU","ASN") & year=="05")
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

#### From the computation above, we can see that the number of genes with q-value less than 0.05 is 560. Here, we can see a dramatic drop in the number of genes with q-value < 0.05 when we fix the year. However, the sample size is much smaller in this latest analysis which means we have less power, as can be seen below:
```{r}
table(sampleInfo$ethnicity[index])
```

###9: To provide a more balanced comparison, we repeat the analysis, but now taking 3 random CEU samples from 2002. Repeat the analysis above, but comparing the ASN from 2005 to three random CEU samples from 2002. Set the seed at 3, set.seed(3). How many genes have q-values < 0.05?
```{r}
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
indexASN = which(ethnicity%in% c("ASN") & year=="05")
indexCEU = which(ethnicity%in% c("CEU") & year=="02")
set.seed(3)
indexCEU<-sample(indexCEU,3)
index<-c(indexASN,indexCEU)
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

#### Here, in this example, we can see that the number of genes that have a q-value less than 0.05 is 3695. 
